lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py

```diff
@@ -247,92 +252,118 @@ def argspec(self):
+            k8s_obj = self._wait_for_vm_state(new_state)
+
+        return changed, k8s_obj
 
     def execute_module(self):
-        # Parse parameters specific for this module:
-        definition = virtdict()
+        # Parse parameters specific to this module:
         ephemeral = self.params.get('ephemeral')
         state = self.params.get('state')
-
-        if not ephemeral:
-            definition['spec']['running'] = state == 'running'
-
-        # Execute the CURD of VM:
-        template = definition if ephemeral else definition['spec']['template']
         kind = 'VirtualMachineInstance' if ephemeral else 'VirtualMachine'
-        dummy, definition = self.construct_vm_definition(kind, definition, template)
-        result = self.execute_crud(kind, definition)
-        changed = result['changed']
-
-        # Manage state of the VM:
-        if state in ['running', 'stopped']:
-            if not self.check_mode:
-                ret = self.manage_state(state)
-                changed = changed or ret
+        _used_params = [name for name in self.params if self.params[name] is not None]
+        # Is 'spec:' getting changed?
+        vm_spec_change = True if set(VM_SPEC_PARAMS).intersection(_used_params) else False
+        changed = False
+        method = ''
+
+        if ephemeral and state not in ('present', 'absent'):
+            raise TypeError("Ephemeral VMs only support 'present' and 'absent' states.")
+
+        self.client = self.get_api_client()
+        self._kind_resource = self.find_supported_resource(kind)
+        k8s_obj = self.get_resource(self._kind_resource)
+        if not self.check_mode and not vm_spec_change and state != 'absent' and not k8s_obj:
+            self.fail("It's impossible to create an empty VM or change state of a non-existent VM.")
+
+        # Changes in VM's spec or any changes to VMIs warrant a full CRUD, the latter because
+        # VMIs don't really have states to manage; they're either present or don't exist
+        # Also check_mode always warrants a CRUD, as that'll produce a sane result
+        if vm_spec_change or state in ['present', 'absent'] or ephemeral or self.check_mode:
+            definition = virtdict()
+            if not ephemeral:
+                definition['spec']['running'] = state == 'running'
+            template = definition if ephemeral else definition['spec']['template']
+            dummy, definition = self.construct_vm_definition(kind, definition, template)
+            result = self.execute_crud(kind, definition)
+            changed = result['changed']
+            k8s_obj = result['result']
+            method = result['method']
+
+        if ephemeral and state == 'present' and self.params.get('wait') and not self.check_mode:
+            # Waiting for state==absent is handled inside execute_crud()
+            k8s_obj = self._wait_for_vmi_running()
+
+        if not ephemeral and state in ['running', 'stopped'] and not self.check_mode:
```

Comment:
I think it make sense to support running state for ephmeral VM, as well as stopped and absent. which would be same

Reply:
@pkliczewski we need a tie breaker.

In my code ephemeral vms (aka VMIs) support only states _present_ (VMI exists) and _absent_ (VMI doesn't exist), while normal/non–ephemeral ones (VM objects) support also _running_ and _stopped_. My reasoning is that if ephemeral vms also supported running/stopped, that would mean adding two aliases: running==present (which makes sense for a VMI), but also stopped==absent.

And the latter I really don't like. Without those aliases, if you want to stop an ephemeral vm, you're forced to do `state: absent`, which makes it clear you're actually deleting it. With the aliases added an operator might attempt `state: stopped` and be surprised that the vm got deleted and not stopped. (My imaginary operator is someone accustomed to traditional virt systems, for whom such behavior is non–obvious and surprising.)

We need someone to call on this. Three options, choose wisely:
1. ephemerals support only present/absent; no aliases for running/stopped
2. ephemerals support present/running/absent, with running==present; no _stopped_ to avoid the nasty surprise
3. ephemerals support present/running/absent/stopped, with running==present and stopped==absent

Reply:
@mmazur I think that Ondra's suggestion is correct. We need to be consistent how we present vm state to the user. At some point users will be confused if they not understand ephemeral vs stateful vm. For ephemeral vms stopping means delete and it is fine.